packages/ra-core/src/form/getFormInitialValues.ts

```diff
@@ -3,28 +3,27 @@ export default function getFormInitialValues(
     defaultValue,
     record
 ) {
-    let finalInitialValues = {
-        ...initialValues,
-        ...record,
-    };
-
     if (typeof defaultValue !== 'undefined') {
         console.warn(
             '"defaultValue" is deprecated, please use "initialValues" instead'
         );
     }
 
-    if (typeof defaultValue === 'object') {
-        finalInitialValues = {
-            ...defaultValue,
-            ...finalInitialValues,
-        };
-    } else if (typeof defaultValue === 'function') {
-        finalInitialValues = {
-            ...defaultValue(record),
-            ...finalInitialValues,
-        };
+    return {
+        ...getValues(defaultValue, record),
+        ...getValues(initialValues, record),
+        ...record,
```

Comment:
We should probably check whether `initialValues` or `defaultValue` are functions before overriding them with the record as we pass it to them as an argument. What do you think @fzaninotto ?